### DISCUSSION TITLE
Custom Strings not having any effect

### DISCUSSION THREAD

**Discussion by NsinghP** [benzino77/tasmocompiler#169] [Q&A]
Hello,
It is my understanding that, the command WifiConfig 7 will prevent WiFiManager from jumping into action and making the tasmota devices available using an unsecured access point in case of Home Wifi network failure, though this feature will remain active for Reset case.

(WifiConfig 7 = set Wi-Fi Manager (web server at 192.168.4.1) as the current configuration tool restricted to reset settings only.)

I am using TasmoCompiler to make a custom .bin file using Wemos/Nodemcu Option and following Custom Strings:

#define USE_UFILESYS
#define GUI_TRASH_FILE
#define MODULE            USER_MODULE
#define USER_TEMPLATE "{\"NAME\":\"MYHOME\",\"GPIO\":[258,259,32,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}"
#define USE_EMULATION_HUE
#define EMULATION         EMUL_HUE
#define DEVICE_NAME       "First Room"  
#define FRIENDLY_NAME     "First Room Switch"
#define TIME_DST_OFFSET    +330          
#define TIME_STD_OFFSET    +330      
#define APP_TIMEZONE       99
#define USE_TASMOTA_DISCOVERY
#define WIFI_CONFIG_TOOL WIFI_MANAGER_RESET_ONLY

But the last two lines don't seem to do anything. 

For line (#define USE_TASMOTA_DISCOVERY) TasmoManager does not detect tasmota devices present on the network and for line (#define WIFI_CONFIG_TOOL WIFI_MANAGER_RESET_ONLY) The default value does not change even after reset 1 command.

Command CMD: wificonfig gives the foloowing result:
RSL: RESULT = {"WifiConfig":{"0":"Restart"}}

But it should be giving the following result:
RSL:RESULT = {"WifiConfig":{"7":"ManagerRst"}}

I have tried to troubleshoot both the issues about (#define USE_TASMOTA_DISCOVERY) in https://github.com/arendst/Tasmota/discussions/11509 and Posted about (#define WIFI_CONFIG_TOOL WIFI_MANAGER_RESET_ONLY) issue in https://github.com/arendst/Tasmota/discussions/11845 .

It looks like both the lines lead to desired results if using VSCode+PlatformIO to compile the .bin but they don't work if using TasmoCompiler. Please confirm if the issue is with TasmoCompiler. 

When launching TasmoCompiler in Firefox i do get following messages: 

warning file-loader@2.0.0: Invalid bin field for "file-loader".
warning mini-css-extract-plugin@0.5.0: Invalid bin field for "mini-css-extract-plugin".
warning url-loader@1.1.2: Invalid bin field for "url-loader".
warning Your current version of Yarn is out of date. The latest version is "1.22.5", while you're on "1.22.4".

I don't know if they are relevent to my issue. 

Regards,

Navi

**Comment by benzino77** [comment]
1. You don't have to define those two lines:
```c
#define USE_UFILESYS
#define GUI_TRASH_FILE
```
you can choose `SD Card/LittleFS` from features list

2. When you are using vscode + platformio (via gitpod or not) you probably compiling latest `development` branch. Could you please check the same branch/version when you are using TasmoCompiler (not the version 9.3.1)?

**Reply by NsinghP** [reply]
Hi, 

I am using Tasmota v. 9.4.1 with TasmoCompiler. I have not tried with vscode + platformio, but barbudor at arrends/tasmota discussion has compiled with vscode + platformio and reported success. I am not sure which version,  of tasmota,  barbudor is using, but I can find out. Now that SD Card/LittleFS option is available, I will remove corresponding lines from my custom strings. Thanks for pointing that out. 

Regards,

Navi

**Reply by benzino77** [reply]
This is what is return from console when binary is compiled by TasmoCompiler with your settings, tasmota version 9.4.0:
```
00:00:12.521 CMD: wificonfig
00:00:12.526 RSL: RESULT = {"WifiConfig":{"7":"ManagerRst"}}
```

**Reply by NsinghP** [reply]
No, I get following:

23:15:37.896 CMD: wificonfig
23:15:37.901 RSL: RESULT = {"WifiConfig":{"0":"Restart"}}

**Reply by NsinghP** [reply]
>     1. You don't have to define those two lines:
> 
> 
> ```c
> #define USE_UFILESYS
> #define GUI_TRASH_FILE
> ```
> 
> you can choose `SD Card/LittleFS` from features list

Ok, I compiled using latest development version on TasmoCompiler selecting SD Card/LittleFS and omitted #define USE_UFILESYS, #define GUI_TRASH_FILEthose. File management system is Visible and workding. 
But 
#define USE_TASMOTA_DISCOVERY
#define WIFI_CONFIG_TOOL WIFI_MANAGER_RESET_ONLY
lines are still not working.

**Reply by benzino77** [reply]
How do you flush new bin file to your wemos? If you are using [Tasmotizer](https://github.com/tasmota/tasmotizer) it does not like boards with non standard 1MB memory. Use esptool or other tool.

**Reply by NsinghP** [reply]
I am simply upgrading, using the upgrade firmware option from Web GUI.

PS. after upgrading i issue reset 1 command.

**Reply by benzino77** [reply]
Latest development on my side compiled with your settings using TasmoCompiler:
```
00:00:50.716 CMD: status 2
00:00:50.722 RSL: STATUS2 = {"StatusFWR":{"Version":"9.4.0.1(tasmota)","BuildDateTime":"2021-04-23T20:02:34","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"374/699"}}
00:00:53.212 CMD: wificonfig
00:00:53.217 RSL: RESULT = {"WifiConfig":{"7":"ManagerRst"}}
```
Upload new bin via esptool.

What I suggest is to do `erase_flash` on your wemos and then upload new bin file.
I just cannot reproduce your case.

**Reply by benzino77** [reply]
Try to compile without this line:
```c
#define USER_TEMPLATE "{\"NAME\":\"MYHOME\",\"GPIO\":[258,259,32,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}"
```

**Reply by benzino77** [reply]
All my tests were performed without this line. If I include this line my wemos starts to behave... very strange.

**Reply by NsinghP** [reply]
Alright, I will do an erase flash, but that is going to take some time as I will have to setup the circuit to flash the chip via serial interface. I will do it over the weekend and update the status here. Thanks for your time and help. 

Regards, 
Navi

**Reply by NsinghP** [reply]
> Try to compile without this line:
> 
> ```c
> #define USER_TEMPLATE "{\"NAME\":\"MYHOME\",\"GPIO\":[258,259,32,163,256,257,1,1,162,161,160,1,1,1],\"FLAG\":0,\"BASE\":18}"
> ```

Strange behaviour may be due to GPIO mappings as I am using RX and TX pins.  I have a custom device and I can use all the Available GPIO pins.

**Reply by benzino77** [reply]
You can define template after flushing. It doesn't have to be defined during compilation time.
Anyway:
1. do erase
2. remove template definition
3. compile
4. flush again via serial/usb

**Reply by NsinghP** [reply]
> 
> You can define template after flushing. It doesn't have to be defined during compilation time.
> Anyway:
> 
>     1. do erase
> 
>     2. remove template definition
> 
>     3. compile
> 
>     4. flush again via serial/usb

Ok, I did steps 2 and 3. Still the problem persists. I will hookup the esp8266 via serial and do Erase/Flash over the weekend and see if that solves the problem. Thanks and Regards, Navi.

**Reply by NsinghP** [reply]
I have done an erase/flash the problem still persists: 

I am still gettings 22:15:09.468 CMD: wificonfig
22:15:09.473 RSL: RESULT = {"WifiConfig":{"0":"Restart"}}

What is bugging me is that even if it is not accepting the custom string I am entering, why is it setting wificonfig to "0" when the default should be "4"

![Screenshot (119)](https://user-images.githubusercontent.com/69470381/115966704-6ffed000-a54c-11eb-94d2-0ae11549c8c8.png)

Command: esptool.py --port COM4 --baud 115200 --after no_reset write_flash --flash_mode dio 0x00000 C:\Users\NPHom\Desktop\ESP8266 FLASH\Tasmota\Custom BIN Files\18th run\firmware.bin --erase-all

esptool.py v2.6
Serial port COM4
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
MAC: 84:cc:a8:ae:e8:63
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Erasing flash (this may take a while)...
Chip erase completed successfully in 10.0s
Flash params set to 0x0240
Compressed 596368 bytes to 426240...
Wrote 596368 bytes (426240 compressed) at 0x00000000 in 37.9 seconds (effective 125.9 kbit/s)...
Hash of data verified.

Leaving...
Staying in bootloader.

Firmware successfully flashed. Unplug/replug or reset device 
to switch back to normal boot mode.

**Reply by benzino77** [reply]
It is something I can't reproduce. Here is the output from compiled Tasmota using TasmoCompiler and your settings:
```
00:00:18.492 CMD: status 2
[11:43:45:548] 00:00:18.498 RSL: STATUS2 = {"StatusFWR":{"Version":"9.4.0.1(tasmota)","BuildDateTime":"2021-04-25T11:37:39","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"374/699"}}
[11:43:50:078] 00:00:23.030 CMD: wificonfig
[11:43:50:083] 00:00:23.035 RSL: RESULT = {"WifiConfig":{"7":"ManagerRst"}}
```

During compilation there is a message that wificonfig has changed from default 4 to `WIFI_MANAGER_RESET_ONLY`:
```
In file included from tasmota/my_user_config.h:953:0,
                 from tasmota/StackThunk_light.cpp:28:
tasmota/user_config_override.h:476:0: warning: "WIFI_CONFIG_TOOL" redefined [enabled by default]
 #define WIFI_CONFIG_TOOL WIFI_MANAGER_RESET_ONLY
 ^
In file included from tasmota/StackThunk_light.cpp:28:0:
tasmota/my_user_config.h:78:0: note: this is the location of the previous definition
 #define WIFI_CONFIG_TOOL       WIFI_RETRY        // [WifiConfig] Default tool if wifi fails to connect (default option: 4 - WIFI_RETRY)
```

I'm attaching you the binary compiled on my station. Uncompress it first if you want to flush it with NodeMCU PyFlasher.
[firmware.bin.gz](https://github.com/benzino77/tasmocompiler/files/6371678/firmware.bin.gz)

**Reply by benzino77** [reply]
Here is the binary build on gitpod using TasmoCompiler and your settings.
```
00:00:07.525 RSL: STATUS2 = {"StatusFWR":{"Version":"9.4.0.1(tasmota)","BuildDateTime":"2021-04-25T09:43:36","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"374/699"}}
[11:51:44:540] 00:00:13.524 CMD: wificonfig
[11:51:44:540] 00:00:13.528 RSL: RESULT = {"WifiConfig":{"7":"ManagerRst"}}
```
[firmware.bin.gz](https://github.com/benzino77/tasmocompiler/files/6371687/firmware.bin.gz)

**Reply by benzino77** [reply]
Compiling using TasmoCompiler on gitpod shows the same message that `WIFI_CONFIG_TOOL` has been changed from default "4":
```
In file included from tasmota/my_user_config.h:953:0,
                 from /tmp/Tasmota/tasmota/tasmota.ino:43:
tasmota/user_config_override.h:477:0: warning: "WIFI_CONFIG_TOOL" redefined [enabled by default]
 #define WIFI_CONFIG_TOOL WIFI_MANAGER_RESET_ONLY
 ^
In file included from /tmp/Tasmota/tasmota/tasmota.ino:43:0:
tasmota/my_user_config.h:78:0: note: this is the location of the previous definition
 #define WIFI_CONFIG_TOOL       WIFI_RETRY        // [WifiConfig] Default tool if wifi fails to connect (default option: 4 - WIFI_RETRY)
```

**Reply by benzino77** [reply]
Updating my Wemos via WebGUI give me the same results (your template is active):
```
00:01:41.255 RSL: STATUS2 = {"StatusFWR":{"Version":"9.4.0.1(tasmota)","BuildDateTime":"2021-04-25T09:43:36","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"335/699"}}
00:01:45.464 CMD: wificonfig
00:01:45.471 RSL: RESULT = {"WifiConfig":{"7":"ManagerRst"}}
```
Performing `RESET 1`:
```
00:00:24.881 CMD: reset 1
00:00:24.885 RSL: RESULT = {"Reset":"Reset and Restarting"}
00:00:25.581 CFG: Use defaults
00:00:26.789 APP: Restarting

ets Jan  8 2013,rst cause:2, boot mode:(3,0)

load 0x4010f000, len 3584, room 16
tail 0
chksum 0xb0
csum 0xb0
v3969889e
~l

00:00:00.001 HDW: ESP8266EX
00:00:00.017 UFS: FlashFS mounted with 2000 kB free
00:00:00.064 CFG: Loaded from File, Count 1
00:00:00.070 QPC: Count 1
00:00:00.076 Project tasmota First Room Version 9.4.0.1(tasmota)-2_7_4_9(2021-04-25T09:43:36)
00:00:00.365 WIF: WifiManager active for 3 minutes
00:00:00.873 HTP: Web server active on tasmota_7C309F-4255 with IP address 192.168.4.1
00:00:06.881 QPC: Reset
00:00:08.234 CMD: status 2
00:00:08.238 RSL: STATUS2 = {"StatusFWR":{"Version":"9.4.0.1(tasmota)","BuildDateTime":"2021-04-25T09:43:36","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"374/699"}}
00:00:10.328 CMD: wificonfig
00:00:10.332 RSL: RESULT = {"WifiConfig":{"7":"ManagerRst"}}
```
Still works as expected. Maybe try with new ESP826 board? Maybe there is something wrong with this one?

**Reply by NsinghP** [reply]
Hey Thanks, For those files. I Flash/Erased with both of them, this time I also selected (DOUT) mode, but the problem persists. I have noticed another strange behavior that may be causing this problem. After Erase/Flash tasmota Wifi Access point asks for wifi credentials, when I enter my network name and passphrase, it is rejected the first time, then I have to reconnect to the the device access point a second time and it accepts the wifi credentials. I think between the first attempt and second attempt it is changing the default wificonfig to "0". There is no other explanation that I can think of that changes the value either from default "4" or from new "7" (if my custom string has been successfully incorporated into the binary) to "0".

Thanks for your help and time, I will let you know if I ever figure this one out. 

![Screenshot (120)](https://user-images.githubusercontent.com/69470381/116000025-28d81400-a60c-11eb-9963-5f38208f566a.png)

Command: esptool.py --port COM4 --baud 115200 --after no_reset write_flash --flash_mode dout 0x00000 C:\Users\NPHom\Desktop\ESP8266 FLASH\Tasmota\Custom BIN Files\23rd run Benzino Desktop\firmware.bin --erase-all

esptool.py v2.6
Serial port COM4
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
MAC: 84:cc:a8:ae:e8:63
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Erasing flash (this may take a while)...
Chip erase completed successfully in 9.7s
Compressed 596368 bytes to 426239...
Wrote 596368 bytes (426239 compressed) at 0x00000000 in 37.9 seconds (effective 125.9 kbit/s)...
Hash of data verified.

Leaving...
Staying in bootloader.

Firmware successfully flashed. Unplug/replug or reset device 
to switch back to normal boot mode.




Regards,

Navi

**Reply by NsinghP** [reply]
I have compiled again TasmoCompiler. (attaching compile messages in .txt file) Problem still persists and wificonfig gives status of "0" instead of "7". I am thinking that it may be Flash corruption issue. Nothing else makes sense. Also in the compile messages there is no mention of USE_TASMOTA_DISCOVERY, though my custom string includes #define Use_Tasmota_Discovery line. 
[Compile messages.txt](https://github.com/benzino77/tasmocompiler/files/6409851/Compile.messages.txt)

**Comment by benzino77** [comment]
3. If you are using VScode+platformio you probably don't change `platformio_override.ini` file to override `board_build.ldscript = eagle.flash.4m2m.ld` - choose the same settings in TasmoCompiler as you are using compiling vanilla Tasmota on gitpod/VScode

**Reply by NsinghP** [reply]
> 
> 
>     1. If you are using VScode+platformio you probably don't change `platformio_override.ini` file to override `board_build.ldscript = eagle.flash.4m2m.ld` - choose the same settings in TasmoCompiler as you are using compiling vanilla Tasmota on gitpod/VScode

I  have not used VScode+platformia before, so there will be a learnning curve, but how do I select to not change platformia_override.ini file to not reflect `board_build.ldscript = eagle.flash.4m2m.ld` in TasmoCompiler?

**Reply by NsinghP** [reply]
Problem Solved. There was a bug in Tasmota 9.4.0 causing this problem, it has now been fixed. Please see https://github.com/arendst/Tasmota/discussions/11845

Thanks for your time in helping solve this bug. 

Regards,

Navi
